### ISSUE TITLE
GoboLinux: Add a recipe file. 

### ISSUE THREAD

**Issue by mraper3** [dylanaraps/neofetch#713]
## Description
Just want to use neofetch on GoboLinux.

#### Neofetch version
latest git clone

Going off a similar issue, here's some uname info:
uname -a:
Linux gobovm 4.9.16-Gobo #0 SMP PREEMPT Mon Mar 20 08:53:55 BRT 2017 x86_64 GNU/Linux
uname -s:
Linux
uname -r:
4.9.16-Gobo
uname -m:
x86_64
uptime:
16:04:54 up 26 min, 4 users, load average: 0.07, 0.04, 0.00
df:
Filesystem        1K-blocks         Used  Available  Use%  Mounted on
/dev/root            51342972     7012580  41692620  15%  /
devtmpfs               1758660               0       1758660    0%   /dev
none                     1759968               0       1759968    0%   /Data/Variable/run/udev
none                     1759968               0       1759968    0%   /dev/shm

(Had to hand-type that instead of copy and paste 'cause their copy of Firefox is still buggy somehow)

Keep in mind that GoboLinux uses a different filesystem structure than standard Linux, but is still compatible with it through use of symlinks, mostly.  They explain it better than I could on their website.

If you need more information about this, just let me know!  Thanks, in advance.

**Comment by mraper3** [comment]
Erm, well, I actually got it installed, and it's even got a logo and everything, so it seems to already support GoboLinux.  So I'll just close this now.  Sorry for wasting your time.

**Comment by konimex** [comment]
Can you do a screenshot of Neofetch in Gobo?

Actually, I did add the detection and package count in Gobo.

**Comment by mraper3** [comment]
I'm reopening this issue because I'd like to establish a package ("Recipe") for neofetch in GoboLinux.  Recipes are basically wrappers around the typical Linux install process that accounts for GoboLinux's differing filesystem structure.

I closed this issue earlier since the way I had installed it was just the usual "make install" (which did actually work), but I figured if that could be done, then it could also be done the way the OS devs intended.

The process doesn't involve modifying your program at all.  GoboLinux's (GL's) Recipe maker program detects the install method (in neofetch's case, Makefile) and runs appropriately.  
Then you Compile it using their Compile program which wraps the "make install" part, and submit the packaged Recipe to their devs for review.

If/when they approve it, it then becomes an official recipe which can be obtained from their (non-GitHub) repo.

See the below files for initial files that I want to submit to the OS devs; these are part of the packaged Recipe.
[Recipe.txt](https://github.com/dylanaraps/neofetch/files/984681/Recipe.txt)
[Dependencies.txt](https://github.com/dylanaraps/neofetch/files/984686/Dependencies.txt)
[Description.txt](https://github.com/dylanaraps/neofetch/files/984687/Description.txt)
Note that these contain just the information GoboLinux needs, not the actual neofetch program.  The Compile program obtains that over the internet via git.

I wanted to run this by the neofetch devs first to see if it was OK with them, and if there was anything further I needed to do, or information I could give about this.

**Comment by dylanaraps** [comment]
The files look good to me.

Is there any way of listing optional dependencies in GoboLinux? If not is there a way to display a post-install message that directs the user to this page?:

https://github.com/dylanaraps/neofetch/wiki/Dependencies

If it isn't possible that's fine too, it's just that we've got a lot of optional dependencies.

**Comment by mraper3** [comment]
Sorry for the info dump below, but here is the relevant information about dependencies and optional dependencies.  Looks like the available "use flags" (dependencies) they list are based solely on the Recipes in their repo, so some of your optionals may or may not be there.  At the end of the info dump is a quick blurb about a PostInstall bash script that can be used; I'll have to dig up more details on it.  If the Use Flags functionality won't work for you, then the PostInstall script probably will.

Begin info dump (this is from their wiki):
-----

BuildDependencies
This file lists dependencies which must be present to successfully compile the Program. They may include compilers or build tools. The format is the same as Dependencies, below.

BuildInformation
Informational file about which versions of dependencies were actually linked against when compiling. 

Dependencies
This file lists programs which should be installed for this Program to work properly. The format is like

    Fontconfig 2.4.2 
    FreeType 2.1.10 
    GCC >= 3.0.0, < 4.0.0, != 3.1.0 
    Glibc 2.5 
    Lame >= 3.96.1 [lame] 
    Mesa 6.5.2 
    Qt >= 3.3.8, < 4.0 [qt] 
    LuaRocks:luafilesystem 
    CPAN:XML::Parser 
    Xorg 7.2 
    ZLib 1.2.3 

The tags such as [lame] specify use flags, optional dependencies which affect the compilation of the package, if present.

The exact algorithm for complex dependencies is specified in CheckDependencies but allows for a sequence of options separated by "|" (or) each of which is a sequence of versions separated by "," (and). Precedence is left to right. Thus:

    GCC < 4.0.0 | >= 4.1.0, != 4.1.2 | ICC > 2.0.0 

means a GCC version less than 4.0.0 or a GCC version greater than 4.1.0 but not equal to 4.1.2 or an ICC version greater than 2.0.0. See the code for CheckDependencies for the exact algorithm.

Dependencies to language-specific package managers can be fulfilled using the Aliens subsystem, using the syntax AlienType:alien_package, as in the examples above.

Note that there are limitations in version handling for Aliens, as it depends on the Alien provider and the package manager itself:

    CPAN dependencies ignore version information (CPAN only installs the latest version)

For more info, see Dependencies and Use Flags. 

Use flags
Use Flags are a way to affect the compilation of a Recipe based on the Programs installed on a system and the user's preferences, without having to edit the Recipe explicitly. Different configure options may be passed to the program being compiled, and additional hook functions may be run by Compile, depending on which flags are activated. Flags may be activated by the presence of installed dependencies, or explicitly by the user.

Also see Available use flags.

Available use flags
    acl 
    avahi # use avahi for bonjour support 
    bonjour 
    curses 
    cyrus_sasl 
    dbus 
    gnome 
    gnome_vfs 
    gnutls # use gnutls for ssl 
    gstreamer 
    gtk 
    gui # enables gui regardless of what toolkit is used. 
    ldap 
    libgnome 
    libgnomeui 
    meanwhile # use meanwhile for sametime support 
    mono 
    networkmanager 
    nss # use nss for ssl 
    ogg 
    openssl # use openssl for ssl 
    pam 
    pango 
    perl 
    qt4 
    rtl #right-to-left text support 
    sametime 
    sqlite 
    ssl 
    startup_notification 
    theora 
    tcl 
    tk 
    x11 # enabled graphical interface 

Use Flags Names
Flags are lower-case alphanumeric plus underscore, and should be named after what they do - the other program, tool, hardware, or functionality they enable, with any other characters stripped out if necessary. Example names would be python, ipw2200, gtk.
Enabling Use Flags

Flags are enabled in three ways. The first is through the global default flag set, in /Programs/Scripts/Current/Data/SystemUseFlags.conf. At the moment, this set is empty, but it will probably end up populated based on what is used for packages and the ISO. The second is the local flags, set in /System/Settings/UseFlags.conf. Finally, the USE environment variable is read for flag specifications. It is intended that this be used for single Compile runs, such as for testing recipes, and not to set flags for your system (use UseFlags.conf for that). Later flag specifications overwrite earlier ones, both in the order listed above and within the files.

A flag specification has the format (-|+)<flag>[ program1[ program2 ...]]. + enables the flag, - disables it, and providing a space-separated list of programs after the flag makes that specification apply only to those programs. Only one flag specification should be included on each line, and everything after a # is ignored as a comment. An example file for clarity:

 +foo # Enable foo globally. This text is ignored.
 -bar
 +bar FooBar

This enables the foo flag globally and disables bar, but then enables the bar flag for only the program FooBar. If the last two lines were the other way around, bar would be globally disabled again. A special specification is -*. This disables all flags, and is probably most useful in the environment variable.

The environment variable takes a space-separated list of flag specifications (rather than newline), so it accepts a special syntax for the specifications, with @ instead of a space when listing programs to go with a flag. It takes the ugly syntax because it's likely to be by far the least common way of using it. The same set of flags from above could be applied with:

 USE="+foo -bar +bar@FooBar"

Use Flags in Recipes

Flags should be listed in the Dependencies or BuildDependencies file in the same manner as the existing cross/!cross flag:

FooBar >= 1.2 [foo,bar]

Flags are separated by commas, and treated as a disjunction - the dependency will be enabled if and only if at least one of the listed flags is enabled. If the cross/!cross flag is specified, it must be the first flag listed (technical limitation; may be lifted in time).

If a flag has no associated dependency, first consider whether it is necessary at all, or whether support should just be enabled by default. If the flag is necessary, it should be listed at the end of the Dependencies file as an entry without a corresponding dependency:

FooBar >= 1.2 [foo,bar]
[baz,quux]

Dependency-free flags may be necessary when associated with hardware, or when there is some lengthy compilation, large filesize, or mutual incompatibility associated with them.

PostInstall

A bash script which is executed by Compile (or InstallPackage) after installation. This is for one-time actions which should not be associated with any stage of the compilation or installation process, but run after the Program is symlinked. They are kept separate from the Recipe file so that they are retained in binary packages which may be distributed. 

-----
End info dump

**Comment by dylanaraps** [comment]
A post install script that just echos the link to the dependencies page
with a short message seems much simpler and easier imo.

**Comment by mraper3** [comment]
Agreed, I'll figure out the particulars on the script and report back.

**Comment by dylanaraps** [comment]
Just a few printfs should suffice:

    printf "%s\n" "Neofetch has many optional dependencies, see this wiki
page for more info:"
    printf "%s\n" "https://GitHub.com/bla"

**Comment by mraper3** [comment]
Alright, I put the above two lines into a bash script named PostInstall.  I used the URL you gave previously, since the later one didn't work.  I'll go ahead and Compile as is, package and submit the recipe to the GoboLinux devs for approval, and keep you posted on how it goes.

**Comment by mraper3** [comment]
Recipe submitted.  It's late here in the States, so I'll probably hear back tomorrow.

**Comment by mraper3** [comment]
I'm still waiting on them to review the Recipe.  It's weird; they've reviewed a few others that were pending at the same time, but not mine.  I'm trying to get in contact with them and see what's going on.

**Comment by mraper3** [comment]
Got in touch with them; I'm just in line for a reviewer.  Dunno when that'll happen, so this might take a while.

**Comment by mraper3** [comment]
The GoboLinux devs approved the Neofetch recipe, but they removed the PostInstall message you asked me to include.  I'll ask them why, but maybe in the meantime I can change the Description file to put the message in there.

**Comment by dylanaraps** [comment]
I'll add something to the man page and help output about the dependencies instead. Though it'd be nice to have both.

**Comment by mraper3** [comment]
Thanks, I'll let you know their response.  Thanks for your patience with this!
